### ISSUE TITLE
Call the 'sk_value' , *(st->data)=0 bad address accessed

### ISSUE THREAD

**Issue by RockySYR** [openssl/openssl#18556]
I need to call the RandomSeed function to generate a random number.  During inserts the structure pointed to by data into table , call 
the function 'getrn' ,get an error at 'hash = (*(lh->hash)) (data)' .The error is   *(st->data)=0, bad address accessed.

Specially marked with  `**<-The error occurs here**` !

**File: lhash.c** 
```
void *lh_insert(_LHASH *lh, void *data)
{
    unsigned long hash;
    LHASH_NODE *nn, **rn;
    void *ret;

    lh->error = 0;
    if (lh->up_load <= (lh->num_items * LH_LOAD_MULT / lh->num_nodes))
        expand(lh);
    rn = getrn(lh, data, &hash);**<-The error occurs here**
    if (*rn == NULL) {
        if ((nn = (LHASH_NODE *)OPENSSL_malloc(sizeof(LHASH_NODE))) == NULL) {
            lh->error++;
            return (NULL);
        }
        nn->data = data;
        nn->next = NULL;
#ifndef OPENSSL_NO_HASH_COMP
        nn->hash = hash;
#endif
        *rn = nn;
        ret = NULL;
        lh->num_insert++;
        lh->num_items++;
    } else {                    /* replace same key */

        ret = (*rn)->data;
        (*rn)->data = data;
        lh->num_replace++;
    }
    return (ret);
}
```
**File: lhash.c** 
```
static LHASH_NODE **getrn(_LHASH *lh, const void *data, unsigned long *rhash)
{
    LHASH_NODE **ret, *n1;
    unsigned long hash, nn;
    LHASH_COMP_FN_TYPE cf;
    hash = (*(lh->hash)) (data);**<-The error occurs here**
    lh->num_hash_calls++;
    *rhash = hash;

    nn = hash % lh->pmax;
    if (nn < lh->p)
        nn = hash % lh->num_alloc_nodes;

    cf = lh->comp;
    ret = &(lh->b[(int)nn]);
    for (n1 = *ret; n1 != NULL; n1 = n1->next) {
#ifndef OPENSSL_NO_HASH_COMP
        lh->num_hash_comps++;
        if (n1->hash != hash) {
            ret = &(n1->next);
            continue;
        }
#endif
        lh->num_comp_calls++;
        if (cf(n1->data, data) == 0)
            break;
        ret = &(n1->next);
    }
    return (ret);
}
```
**File: o_names.c** 
```
static unsigned long obj_name_hash(const void *a_void)
{
    unsigned long ret;
    const OBJ_NAME *a = (const OBJ_NAME *)a_void;
    if ((name_funcs_stack != NULL)
        && (sk_NAME_FUNCS_num(name_funcs_stack) > a->type)) {
            
        ret =
            sk_NAME_FUNCS_value(name_funcs_stack,
                                a->type)->hash_func(a->name);**<-The error occurs here**
    } else {
        ret = lh_strhash(a->name);
    }
    ret ^= a->type;
    return (ret);
}
```
**File: stack.c** 
```
void *sk_value(const _STACK *st, int i)
{   
    int index;
    if (!st || (i < 0) || (i >= st->num)){
        return NULL;
        }
    return st->data[i];**<-The error occurs here**
}
```

The 'st->data' = 0 ,but I can't find when to assign a value to name_funcs_stack. Please give me some help！

**Comment by t8m** [comment]
Unfortunately it is completely unclear to me what you're trying to achieve. Can you possibly paste some of your application code?

**Comment by RockySYR** [comment]
I want to generate a 32 byte salt value，but the program is stuck in the sk_value function when return st->data[i].  
Below is my application code@t8m 
```
GenerateSalt (
  IN OUT UINT8  *SaltValue,
  IN UINTN         SaltSize
  )
{
  if (SaltValue == NULL) {
    return FALSE;
  }
  SaltValue =ZeroMem(SaltValue,SaltSize);
  RandomSeed(NULL, 0);
  RandomBytes(SaltValue, SaltSize);
}
```

**Comment by t8m** [comment]
RandomSeed and RandomBytes are not OpenSSL functions so I still have no idea what they are doing.

**Comment by RockySYR** [comment]
sorry my code didn't paste full.  @t8m 
```
CONST UINT8  DefaultSeed[] = "default seed";
RandomSeed (
  IN  CONST  UINT8  *Seed  OPTIONAL,
  IN  UINTN         SeedSize
  )
{
  if (SeedSize > INT_MAX) {
    return FALSE;
  }
  //
  // The software PRNG implementation built in OpenSSL depends on message digest algorithm.
  // Make sure SHA-1 digest algorithm is available here.
  //
  if (EVP_add_digest (EVP_sha1 ()) == 0) {
    return FALSE;
  }
  //
  // Seed the pseudorandom number generator with user-supplied value.
  // NOTE: A cryptographic PRNG must be seeded with unpredictable data.
  //
  if (Seed != NULL) {
    RAND_seed (Seed, (UINT32) SeedSize);
  } else {
    RAND_seed (DefaultSeed, sizeof (DefaultSeed));
  }
  if (RAND_status () == 1) {
    return TRUE;
  }
  return FALSE;
}

RandomBytes (
  OUT  UINT8  *Output,
  IN   UINTN  Size
  )
{
  //
  // Check input parameters.
  //
  if (Output == NULL || Size > INT_MAX) {
    return FALSE;
  }
  //
  // Generate random data.
  //
  if (RAND_bytes (Output, (UINT32) Size) != 1) {
    return FALSE;
  }
  return TRUE;
}
```

**Comment by RockySYR** [comment]
the program is stuck in the sk_value function when return st->data[i]
```
void *sk_value(const _STACK *st, int i)
{   
    int index;
    if (!st || (i < 0) || (i >= st->num)){
        return NULL;
        }
    return st->data[i];**<-The error occurs here**
}
```

**Comment by t8m** [comment]
It's unclear to me how something can be stuck in return? Do you mean instead that the EVP_add_digest() returns 0? Which openssl release do you use? The EVP_add_digest() should be unnecessary, also with any supported releases it is actually completely useless, because they do not use random number generators based on SHA1 and the openssl implicitly initializes.

**Comment by RockySYR** [comment]
My  library is running in UEFI environment，EVP_add_digest() unnecessary too? Do I need to actively initialize it?
The address of st->data is 0
 I use version 1.0.2

**Comment by RockySYR** [comment]
My  library is running in UEFI environment，EVP_add_digest() unnecessary too? Do I need to actively initialize it?
The address of st->data is 0
 I use version 1.0.2

**Comment by t8m** [comment]
Ah, 1.0.2 is different in this regard. It is also unsupported.

**Comment by t8m** [comment]
Instead of EVP_add_digest(), could you try initializing OpenSSL at the start with `OpenSSL_add_all_algorithms()`?

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of 3.4 dev barring further input
